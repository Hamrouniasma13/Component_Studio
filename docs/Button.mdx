---
name: Button
route: /button
menu: ⚛️ Atoms
---

import { Playground, Props } from 'docz'
import { Button } from '../src'

# Button

The button is the main element of the website.
All props that are not for the button are passed to the [`Link`](/link) used to create the link.

For a lighter button, see [`Simple Button`](/simple-button)

<Button href="/">Click me</Button>

<Props of={Button} />

## Colors

You can choose the color of the button by using the correct key from [theme](/theme#colors).

<Playground>
  <Button href="/" color="primary">
    Primary
  </Button>{' '}
  <Button href="/" color="secondary">
    Secondary
  </Button>{' '}
  <Button href="/" color="third">
    Third
  </Button>{' '}
  <Button href="/" color="fourth">
    Fourth
  </Button>{' '}
  <Button href="/" color="fifth">
    Fifth
  </Button>{' '}
  <Button href="/" color="sixth">
    Sixth
  </Button>
</Playground>

## Sizes variants

You can choose the size of the button, with the `size` prop.
Available options are `big` and `small`.

<Playground>
  <Button href="/" size="big">
    Click me
  </Button>{' '}
  <Button href="/" size="small">
    Click me
  </Button>
</Playground>

## Ghost variant

To display a secondary action, you can use the `ghost` prop.

<Playground>
  <Button href="/" ghost>
    Click me
  </Button>
</Playground>

## Icon

You can display an icon inside the button, with `icon` prop.
See the list of available icons [here](/icon).
You can place it before the label with `iconPosition="left"`

<Playground>
  <Button href="/" icon="cog" />{' '}
  <Button href="/" icon="cog">
    Click me
  </Button>{' '}
  <Button href="/" icon="cog" iconPosition="left">
    Click me
  </Button>
</Playground>

## Real button

If you don't want to display a link, but just a button, don't pass `href`.
Like so, we use `button` instead of [`Link`](/link)

<Playground>
  <Button onClick={() => window.alert('click')}>Click me</Button>
</Playground>
