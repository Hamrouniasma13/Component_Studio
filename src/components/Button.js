/** @jsx jsx */
import { jsx, ThemeContext } from '@emotion/core'
import React from 'react'
import PropTypes from 'prop-types'
import Icon from './Icon'
import Link from './Link'
import { textColor, darken } from '../theme/utils'

/**
 * Calculate padding from size
 * @param {string} size
 */
const getPadding = size => (size === 'big' ? 15 : size === 'small' ? 7 : 10)

function StyledButton({
  button,
  color = 'primary',
  padding,
  ghost,
  css,
  ...props
}) {
  const theme = React.useContext(ThemeContext)
  const mainColor = theme.colors[color]
  const Tag = button ? 'button' : Link

  return (
    <Tag
      {...props}
      css={{
        // Block
        display: 'inline-block',
        position: 'relative',
        verticalAlign: 'middle',
        maxWidth: '700px',
        overflow: 'hidden',
        backgroundColor: ghost ? 'white' : mainColor,
        boxSizing: 'border-box',

        // Border
        padding,
        borderRadius: '4px',
        borderColor: ghost ? '#f3f3f3' : 'transparent',
        borderWidth: '1px',
        borderStyle: 'solid',

        // Text
        textAlign: 'center',
        fontFamily: 'Source Sans Pro, Arial, Helvetica, sans-serif',
        fontWeight: '600',
        fontSize: '14px',
        letterSpacing: '1.1px',
        lineHeight: 1,
        color: ghost ? 'black' : textColor(darken(mainColor, 50), theme),

        // Reset
        outline: 'none',
        textDecoration: 'none',
        zIndex: 0,

        // Optimisation
        textRendering: 'optimizeLegibility',
        WebkitFontSmoothing: 'antialiased',
        MozOsxFontSmoothing: 'grayscale',

        // Other
        cursor: 'pointer',

        // Animation
        transition: 'background-color 0.2s, color 0.2s, border-color 0.2s',
        '&:hover': {
          backgroundColor: ghost ? darken('white', 5) : darken(mainColor, 10),
          color: ghost ? '#555555' : textColor(darken(mainColor, 50), theme),
        },

        // Responsive
        [theme.media.tablet]: {
          width: 'auto',
        },

        ...css,
      }}
    />
  )
}

function StyledIcon({ name, left, right }) {
  return (
    <Icon
      name={name}
      size={20}
      css={{
        verticalAlign: 'bottom',
        marginLeft: right ? 15 : 0,
        marginRight: left ? 15 : 0,
      }}
    />
  )
}

function Label({ children }) {
  return <span css={{ lineHeight: '20px' }}>{children}</span>
}

export default function Button({
  label,
  children,
  icon,
  iconPosition,
  size,
  ...props
}) {
  // deprecate label prop
  const iconOnly = !(children || label)

  // Get content order from icon, children, iconPosition
  const content = [
    icon && iconPosition === 'left' ? (
      <StyledIcon name={icon} left={!iconOnly} key="left-icon" />
    ) : null,
    !iconOnly ? <Label key="label">{children || label}</Label> : null,
    icon && iconPosition !== 'left' ? (
      <StyledIcon name={icon} right={!iconOnly} key="right-icon" />
    ) : null,
  ]

  // Calculate padding from size
  const padding = !iconOnly ? `${getPadding(size)}px 20px` : getPadding(size)

  return (
    <StyledButton button={!props.href} padding={padding} {...props}>
      {content}
    </StyledButton>
  )
}

Button.defaultProps = {
  iconPosition: 'right',
  size: 'normal',
}
Button.propTypes = {
  /** salut descrip */
  icon: PropTypes.string,
  iconPosition: PropTypes.oneOf(['right', 'left']),
  label: PropTypes.string,
  size: PropTypes.oneOf(['small', 'normal', 'big']),
}
