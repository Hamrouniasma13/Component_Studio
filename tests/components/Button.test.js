import React from 'react';
import renderer from 'react-test-renderer';

// wrapper to make our components work with <Link />
import { MemoryRouter } from 'react-router';

// create matcher in order to access to component's style
import { createMatchers } from 'jest-emotion';
import * as emotion from 'emotion';

import Button from '../../components/Button';

expect.extend(createMatchers(emotion));

describe('Component: <Button />', () => {
  it('render label', () => {
    const tree = renderer.create(<Button>See more</Button>);
    expect(tree).toMatchSnapshot();
  });

  it('render with "href"', () => {
    const tree = renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <Button href="/">See more</Button>
      </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
  });

  it('render with "href" + "icon"', () => {
    const tree = renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <Button href="/" icon="close">
          See more
        </Button>
      </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
  });

  it('render with no action button', () => {
    const tree = renderer.create(<Button icon="close">See more</Button>);

    expect(tree).toMatchSnapshot();
  });
});
