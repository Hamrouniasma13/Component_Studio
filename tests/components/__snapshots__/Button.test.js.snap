// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Button /> render "simple" + "icon" 1`] = `
.emotion-1 {
  cursor: pointer;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: bold;
  color: primary;
  font-family: Source Sans Pro,Arial,Helvetica,sans-serif;
  font-size: 13px;
  -webkit-transition: color 0.3 ease;
  transition: color 0.3 ease;
  display: inline-block;
}

.emotion-1:hover {
  color: #2B30F2;
}

.emotion-0 {
  color: inherit;
  margin-right: 10px;
  vertical-align: text-bottom;
}

<button
  className="Button  emotion-1"
  color="primary"
  onClick={[Function]}
  size="normal"
  type="button"
>
  <div
    className="emotion-0"
    colorindex={undefined}
    fill="currentColor"
    size={20}
    style={
      Object {
        "alignItems": "center",
        "display": "inline-flex",
        "justifyContent": "center",
      }
    }
  >
    <svg
      fill="currentColor"
      height={20}
      style={
        Object {
          "display": "inline-block",
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 24 24"
      width={20}
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
  See more
</button>
`;

exports[`Component: <Button /> render "simple" 1`] = `
.emotion-0 {
  cursor: pointer;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: bold;
  color: primary;
  font-family: Source Sans Pro,Arial,Helvetica,sans-serif;
  font-size: 13px;
  -webkit-transition: color 0.3 ease;
  transition: color 0.3 ease;
  display: inline-block;
}

.emotion-0:hover {
  color: #2B30F2;
}

<button
  className="Button  emotion-0"
  color="primary"
  onClick={[Function]}
  size="normal"
  type="button"
>
  See more
</button>
`;

exports[`Component: <Button /> render label 1`] = `
.emotion-0 {
  cursor: pointer;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: bold;
  background-color: #494df4;
  color: #ffffff;
  font-family: Source Sans Pro,Arial,Helvetica,sans-serif;
  font-size: 15px;
  border-radius: 4px;
  -webkit-letter-spacing: 1.1px;
  -moz-letter-spacing: 1.1px;
  -ms-letter-spacing: 1.1px;
  letter-spacing: 1.1px;
  padding: 0 20px;
  position: relative;
  z-index: 0;
  overflow: hidden;
  line-height: 40px;
  text-align: center;
  max-width: 700px;
  vertical-align: middle;
  display: inline-block;
  border: none;
}

.emotion-0:after {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  background-color: #000014;
  content: "";
  height: 530px;
  left: 50%;
  opacity: 0;
  position: absolute;
  top: 50%;
  -webkit-transform: translate3d(-50%,-50%,0) rotate(45deg);
  -ms-transform: translate3d(-50%,-50%,0) rotate(45deg);
  transform: translate3d(-50%,-50%,0) rotate(45deg);
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  width: 0;
  z-index: -1;
}

.emotion-0:hover:after {
  opacity: 1;
  -webkit-transform: translate3d(-50%,-50%,0) rotate(45deg);
  -ms-transform: translate3d(-50%,-50%,0) rotate(45deg);
  transform: translate3d(-50%,-50%,0) rotate(45deg);
  width: 100%;
}

<button
  className="Button  emotion-0"
  color="primary"
  onClick={[Function]}
  size="normal"
  type="button"
>
  See more
</button>
`;

exports[`Component: <Button /> render with "simple" + "to" + "icon" 1`] = `
.emotion-1 {
  cursor: pointer;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: bold;
  color: primary;
  font-family: Source Sans Pro,Arial,Helvetica,sans-serif;
  font-size: 13px;
  -webkit-transition: color 0.3 ease;
  transition: color 0.3 ease;
  display: inline-block;
}

.emotion-1:hover {
  color: #2B30F2;
}

.emotion-0 {
  color: inherit;
  margin-right: 10px;
  vertical-align: text-bottom;
}

<a
  aria-current="page"
  className="Button  emotion-1 active"
  color="primary"
  href="/"
  onClick={[Function]}
  size="normal"
  style={Object {}}
>
  <div
    className="emotion-0"
    colorindex={undefined}
    fill="currentColor"
    size={20}
    style={
      Object {
        "alignItems": "center",
        "display": "inline-flex",
        "justifyContent": "center",
      }
    }
  >
    <svg
      fill="currentColor"
      height={20}
      style={
        Object {
          "display": "inline-block",
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 24 24"
      width={20}
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
  See more
</a>
`;

exports[`Component: <Button /> render with "simple" + "to" 1`] = `
.emotion-0 {
  cursor: pointer;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: bold;
  color: primary;
  font-family: Source Sans Pro,Arial,Helvetica,sans-serif;
  font-size: 13px;
  -webkit-transition: color 0.3 ease;
  transition: color 0.3 ease;
  display: inline-block;
}

.emotion-0:hover {
  color: #2B30F2;
}

<a
  aria-current="page"
  className="Button  emotion-0 active"
  color="primary"
  href="/"
  onClick={[Function]}
  size="normal"
  style={Object {}}
>
  See more
</a>
`;

exports[`Component: <Button /> render with "to" + "icon" 1`] = `
.emotion-1 {
  cursor: pointer;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: bold;
  background-color: #494df4;
  color: #ffffff;
  font-family: Source Sans Pro,Arial,Helvetica,sans-serif;
  font-size: 15px;
  border-radius: 4px;
  -webkit-letter-spacing: 1.1px;
  -moz-letter-spacing: 1.1px;
  -ms-letter-spacing: 1.1px;
  letter-spacing: 1.1px;
  padding: 0 20px;
  position: relative;
  z-index: 0;
  overflow: hidden;
  line-height: 40px;
  text-align: center;
  max-width: 700px;
  vertical-align: middle;
  display: inline-block;
  border: none;
}

.emotion-1:after {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  background-color: #000014;
  content: "";
  height: 530px;
  left: 50%;
  opacity: 0;
  position: absolute;
  top: 50%;
  -webkit-transform: translate3d(-50%,-50%,0) rotate(45deg);
  -ms-transform: translate3d(-50%,-50%,0) rotate(45deg);
  transform: translate3d(-50%,-50%,0) rotate(45deg);
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  width: 0;
  z-index: -1;
}

.emotion-1:hover:after {
  opacity: 1;
  -webkit-transform: translate3d(-50%,-50%,0) rotate(45deg);
  -ms-transform: translate3d(-50%,-50%,0) rotate(45deg);
  transform: translate3d(-50%,-50%,0) rotate(45deg);
  width: 100%;
}

.emotion-0 {
  color: inherit;
  margin-left: 20px;
  vertical-align: text-bottom;
}

<a
  aria-current="page"
  className="Button  emotion-1 active"
  color="primary"
  href="/"
  onClick={[Function]}
  size="normal"
  style={Object {}}
>
  See more
  <div
    className="emotion-0"
    colorindex={undefined}
    fill="currentColor"
    size={20}
    style={
      Object {
        "alignItems": "center",
        "display": "inline-flex",
        "justifyContent": "center",
      }
    }
  >
    <svg
      fill="currentColor"
      height={20}
      style={
        Object {
          "display": "inline-block",
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 24 24"
      width={20}
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
</a>
`;

exports[`Component: <Button /> render with "to" 1`] = `
.emotion-0 {
  cursor: pointer;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: bold;
  background-color: #494df4;
  color: #ffffff;
  font-family: Source Sans Pro,Arial,Helvetica,sans-serif;
  font-size: 15px;
  border-radius: 4px;
  -webkit-letter-spacing: 1.1px;
  -moz-letter-spacing: 1.1px;
  -ms-letter-spacing: 1.1px;
  letter-spacing: 1.1px;
  padding: 0 20px;
  position: relative;
  z-index: 0;
  overflow: hidden;
  line-height: 40px;
  text-align: center;
  max-width: 700px;
  vertical-align: middle;
  display: inline-block;
  border: none;
}

.emotion-0:after {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  background-color: #000014;
  content: "";
  height: 530px;
  left: 50%;
  opacity: 0;
  position: absolute;
  top: 50%;
  -webkit-transform: translate3d(-50%,-50%,0) rotate(45deg);
  -ms-transform: translate3d(-50%,-50%,0) rotate(45deg);
  transform: translate3d(-50%,-50%,0) rotate(45deg);
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  width: 0;
  z-index: -1;
}

.emotion-0:hover:after {
  opacity: 1;
  -webkit-transform: translate3d(-50%,-50%,0) rotate(45deg);
  -ms-transform: translate3d(-50%,-50%,0) rotate(45deg);
  transform: translate3d(-50%,-50%,0) rotate(45deg);
  width: 100%;
}

<a
  aria-current="page"
  className="Button  emotion-0 active"
  color="primary"
  href="/"
  onClick={[Function]}
  size="normal"
  style={Object {}}
>
  See more
</a>
`;

exports[`Component: <Button /> render with no action button 1`] = `
.emotion-1 {
  cursor: pointer;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  font-weight: bold;
  background-color: #494df4;
  color: #ffffff;
  font-family: Source Sans Pro,Arial,Helvetica,sans-serif;
  font-size: 15px;
  border-radius: 4px;
  -webkit-letter-spacing: 1.1px;
  -moz-letter-spacing: 1.1px;
  -ms-letter-spacing: 1.1px;
  letter-spacing: 1.1px;
  padding: 0 20px;
  position: relative;
  z-index: 0;
  overflow: hidden;
  line-height: 40px;
  text-align: center;
  max-width: 700px;
  vertical-align: middle;
  display: inline-block;
  border: none;
}

.emotion-1:after {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  background-color: #000014;
  content: "";
  height: 530px;
  left: 50%;
  opacity: 0;
  position: absolute;
  top: 50%;
  -webkit-transform: translate3d(-50%,-50%,0) rotate(45deg);
  -ms-transform: translate3d(-50%,-50%,0) rotate(45deg);
  transform: translate3d(-50%,-50%,0) rotate(45deg);
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  width: 0;
  z-index: -1;
}

.emotion-1:hover:after {
  opacity: 1;
  -webkit-transform: translate3d(-50%,-50%,0) rotate(45deg);
  -ms-transform: translate3d(-50%,-50%,0) rotate(45deg);
  transform: translate3d(-50%,-50%,0) rotate(45deg);
  width: 100%;
}

.emotion-0 {
  color: inherit;
  margin-left: 20px;
  vertical-align: text-bottom;
}

<button
  className="Button  emotion-1"
  color="primary"
  onClick={[Function]}
  size="normal"
  type="button"
>
  See more
  <div
    className="emotion-0"
    colorindex={undefined}
    fill="currentColor"
    size={20}
    style={
      Object {
        "alignItems": "center",
        "display": "inline-flex",
        "justifyContent": "center",
      }
    }
  >
    <svg
      fill="currentColor"
      height={20}
      style={
        Object {
          "display": "inline-block",
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 24 24"
      width={20}
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
</button>
`;
